Migrate to modern bundler & React 18

echocrow opened the pull request:
Continuation of #743.
Closes #652.

Continuing from the previous PR:

Short-term cleanup:
- [x] Replace legacy `@reach/router` syntax to newer `react-router-dom` syntax.
  (pure refactor)
  (@kasperkristensen noticed you had already started working on this, cheers!)

Long-term cleanup:
- [ ] (optional) Remove `gatsby` and `@reach/router` compatibility shims.
  (leaving them for a bit shouldn't hurt; they are light and are just there to ease migration for those who may have extended the repo but haven't switched their imports & syntax to `react-router-dom` yet)
  Are we okay keeping these in for a few weeks/months after merge? Or would you rather burn these bridges entirely prior to merging?

echocrow commented:
one bug I noticed while replacing the legacy router: duplicating existing discounts was busted.

Going through the code, I believe this is due to a version discrepancy between API server and admin client on my end, so this may be a non-issue. - Looks like the admin client expects a `regions` array on the discount object, but the server is not including this in the discounts API. I suspect this was a recent addition, that did not yet make it into the `medusa` package yet? (I tried to patch my dev backend via `medusa-dev`, but am unsure if that stuck.)

kasperkristensen commented:
> one bug I noticed while replacing the legacy router: duplicating existing discounts was busted.
> 
> Going through the code, I believe this is due to a version discrepancy between API server and admin client on my end, so this may be a non-issue. - Looks like the admin client expects a `regions` array on the discount object, but the server is not including this in the discounts API. I suspect this was a recent addition, that did not yet make it into the `medusa` package yet? (I tried to patch my dev backend via `medusa-dev`, but am unsure if that stuck.)

Thanks for pointing it out! I fixed it in the PR to merge `next -> develop`.

The PR looks great to me, just need to resolve the conflicts, and then it LGTM 🚀

echocrow commented:
> Thanks for pointing it out! I fixed it in the PR to merge next -> develop.
> 
> The PR looks great to me, just need to resolve the conflicts, and then it LGTM 🚀

odd, I think I rebased before opening. 🤔 guessing these files were changed shortly afterwards.

if nobody beats me to it, I'll look into them this weekend and get them resolved.

kasperkristensen commented:
All good @echocrow I removed the final @reach/router usages and the dependency in a separate PR, as we are aiming to release your changes either today or the start of next week. If I missed something then feel free to open a new PR.

echocrow commented:
Understood! Glad to see movement despite me being preoccupied.
Just saw this was even moved into main! 🚀

Will open a new PR re router if that's cool, as—minor conflicts aside—the transition should be pretty much done.

echocrow commented:
[nvm](https://github.com/medusajs/admin/pull/802) - saw this was already done
